### DISCUSSION TITLE
libvalhalla_service_test.so" is for EM_X86_64 (62) instead of EM_AARCH64 (183)

### DISCUSSION THREAD

**Discussion by amanp6323** [valhalla/valhalla#3400] [Q&A]
Getting this error while running .so for valhalla library on android studio platform so plz give any suggestions
java.lang.UnsatisfiedLinkError: dlopen failed: "/data/app/com.example.myhellojni-Wn-7pSsH9gurfMA9UZQ7Ow==/lib/arm64/libvalhalla_service_test.so" is for EM_X86_64 (62) instead of EM_AARCH64 (183)
        at java.lang.Runtime.loadLibrary0(Runtime.java:1071)
        at java.lang.Runtime.loadLibrary0(Runtime.java:1007)
        at java.lang.System.loadLibrary(System.java:1667)
        at com.example.myhellojni.MainActivity.<clinit>(MainActivity.java:19)

**Comment by amanp6323** [comment]
we are creating .so for following platform
ndk {
            abiFilters 'armeabi-v7a', 'arm64-v8a', 'x86', 'x86_64'
       }

**Comment by kevinkreiser** [comment]
sorry there isnt nearly enough information here to help you debug this issue. from the text above it essentially says you built it for `x86_64` but are trying to load it on an `arm64-v8a` based system

**Reply by singhisking11** [reply]
we have created .so via ndk-build for Valhalla library on Ubuntu 20.04 platform and trying to run this .so on android studio for start the Valhalla web server but we are getting the same error.

**Reply by singhisking11** [reply]
And we are getting one more error regarding no implementation found for the methods.

**Reply by singhisking11** [reply]
Do you have any idea about how to load prebuilt .so on android os?

**Reply by kevinkreiser** [reply]
well the error above seems to show that you have build the library for x86_64, you probably dont want that architecture on your android target. instead you want to use the ndk to build the library for one of the arm architectures and link that one in your android target. we've recently had others working on similar things if you look around in the issues y ou will find some clues and even some shared recipes for building this kind of thing. long term we want to make a CI target that builds a minimal android app that can call valhallas native library but we havent had time to get to that

**Reply by singhisking11** [reply]
we have created .so file for odin component of Valhalla and trying to load this on android studio via System.load.library(); function but getting error like no implementation found for the methods of odin.
And we are getting above mentioned error also.
Do you have any idea about it?

**Comment by singhisking11** [comment]
Getting this error while running .so for valhalla library on android studio platform so plz give any suggestions
java.lang.UnsatisfiedLinkError: dlopen failed: "/data/app/com.example.myhellojni-Wn-7pSsH9gurfMA9UZQ7Ow==/lib/arm64/libvalhalla_service_test.so" is for EM_X86_64 (62) instead of EM_AARCH64 (183)
at java.lang.Runtime.loadLibrary0(Runtime.java:1071)
at java.lang.Runtime.loadLibrary0(Runtime.java:1007)
at java.lang.System.loadLibrary(System.java:1667)
at com.example.myhellojni.MainActivity.(MainActivity.java:19)

**Comment by singhisking11** [comment]


**Reply by nilsnolde** [reply]
I could dive into a serious rant right now, but I'll keep it short: it's **your** responsibility to get things working, and you're relying here on free advice from highly experienced and expensive engineers. act accordingly!

**Comment by SparkStar21** [comment]
is for EM_X86_64 (62) instead of EM_AARCH64 (183)
